File: messageDotNet/grpc/baseData/player.cs
using System;
using UnityEngine;


[Serializable]
public class GrpcPlayerFeature
{
    public int Eyebrow;
    public int Mouth;
    public int Eye;
    public int Face;
    public int Hair;
    public int Glove;
    public int Clothes;
    public int Pants;

    public void Set(GameMessageCore.PlayerFeature feature)
    {
        if (feature == null)
        {
            return;
        }
        Eyebrow = feature.Eyebrow;
        Mouth = feature.Mouth;
        Eye = feature.Eye;
        Face = feature.Face;
        Hair = feature.Hair;
        Glove = feature.Glove;
        Clothes = feature.Clothes;
        Pants = feature.Pants;
    }

    public GameMessageCore.PlayerFeature ToProtoData()
    {
        return new GameMessageCore.PlayerFeature()
        {
            Eyebrow = Eyebrow,
            Mouth = Mouth,
            Eye = Eye,
            Face = Face,
            Hair = Hair,
            Glove = Glove,
            Clothes = Clothes,
            Pants = Pants,
        };
    }
}


// 对应 proto.PlayerBaseData
[Serializable]
public class GrpcPlayerBaseData
{
    public long UserId;
    public string Name;
    public int RoleId;
    public string Gender;
    public string RoleIcon;
    public GrpcPlayerFeature Feature;
    public bool Guide;
    public void Set(GameMessageCore.PlayerBaseData baseData)
    {
        if (baseData == null)
        {
            return;
        }
        UserId = baseData.UserId;
        Name = baseData.Name;
        RoleId = baseData.RoleId;
        Gender = baseData.Gender;
        RoleIcon = baseData.RoleIcon;
        Guide = baseData.Guide;
        Feature.Set(baseData.Feature);
    }

    public GameMessageCore.PlayerBaseData ToProtoData()
    {
        return new GameMessageCore.PlayerBaseData
        {
            UserId = UserId,
            Name = Name,
            RoleId = RoleId,
            Gender = Gender,
            RoleIcon = RoleIcon,
            Feature = Feature.ToProtoData(),
            Guide = Guide,
        };
    }

}

File: messageGo/grpc/baseData/player.go
package base_data

import "game-message-core/proto"

type GrpcPlayerFeature struct {
	Eyebrow int32 `json:"eyebrow"`
	Mouth   int32 `json:"mouth"`
	Eye     int32 `json:"eye"`
	Face    int32 `json:"face"`
	Hair    int32 `json:"hair"`
	Glove   int32 `json:"glove"`
	Clothes int32 `json:"clothes"`
	Pants   int32 `json:"pants"`
	Skin    int32 `json:"skin"`
}

func (p *GrpcPlayerFeature) Clear() {
	p.Eyebrow = 0
	p.Mouth = 0
	p.Eye = 0
	p.Face = 0
	p.Hair = 0
	p.Glove = 0
	p.Clothes = 0
	p.Pants = 0
	p.Skin = 0
}

func (p *GrpcPlayerFeature) Set(feature *proto.PlayerFeature) {
	if feature == nil {
		return
	}
	p.Eyebrow = feature.Eyebrow
	p.Mouth = feature.Mouth
	p.Eye = feature.Eye
	p.Face = feature.Face
	p.Hair = feature.Hair
	p.Glove = feature.Glove
	p.Clothes = feature.Clothes
	p.Pants = feature.Pants
	p.Skin = feature.Skin
}
func (p *GrpcPlayerFeature) ToProtoData() *proto.PlayerFeature {
	return &proto.PlayerFeature{
		Eyebrow: p.Eyebrow,
		Mouth:   p.Mouth,
		Eye:     p.Eye,
		Face:    p.Face,
		Hair:    p.Hair,
		Glove:   p.Glove,
		Clothes: p.Clothes,
		Pants:   p.Pants,
		Skin:    p.Skin,
	}
}

// 对应 proto.PlayerBaseData
type GrpcPlayerBaseData struct {
	UserId   int64             `json:"userId"`
	Name     string            `json:"name"`
	RoleId   int32             `json:"roleId"`
	Gender   string            `json:"gender"`
	RoleIcon string            `json:"roleIcon"`
	Feature  GrpcPlayerFeature `json:"feature"`
	Guide    bool              `json:"guide"`
}

func (p *GrpcPlayerBaseData) Clear() {
	p.UserId = 0
	p.Name = ""
	p.RoleId = 0
	p.Gender = ""
	p.RoleIcon = ""
	p.Guide = false
	p.Feature.Clear()
}

func (p *GrpcPlayerBaseData) Set(baseData *proto.PlayerBaseData) {
	if baseData == nil {
		return
	}
	p.UserId = baseData.UserId
	p.Name = baseData.Name
	p.RoleId = baseData.RoleId
	p.Gender = baseData.Gender
	p.RoleIcon = baseData.RoleIcon
	p.Guide = baseData.Guide
	p.Feature.Set(baseData.Feature)
}
func (p *GrpcPlayerBaseData) ToProtoData() *proto.PlayerBaseData {
	pbData := &proto.PlayerBaseData{
		UserId:   p.UserId,
		Name:     p.Name,
		RoleId:   p.RoleId,
		Gender:   p.Gender,
		RoleIcon: p.RoleIcon,
		Feature:  p.Feature.ToProtoData(),
		Guide:    p.Guide,
	}

	return pbData
}
